main/src/ca/uwaterloo/flix/language/phase/unification/BddFormula.scala

```diff
@@ -15,10 +15,166 @@
+      */
+    override def mkVar(id: Int): BddFormula = {
+      new BddFormula(creator.makeIthVar(id))
+    }
+
+    /**
+      * Returns a representation of the complement of `f`.
+      */
+    override def mkNot(f: BddFormula): BddFormula = {
+      new BddFormula(creator.makeNot(f.dd))
+    }
+
+    /**
+      * Returns a representation of the disjunction of `f1` and `f2`.
+      */
+    override def mkOr(f1: BddFormula, f2: BddFormula): BddFormula = {
+      new BddFormula(creator.makeOr(f1.dd, f2.dd))
+    }
+
+    /**
+      * Returns a representation of the conjunction of `f1` and `f2`.
+      */
+    override def mkAnd(f1: BddFormula, f2: BddFormula): BddFormula = {
+      new BddFormula(creator.makeAnd(f1.dd, f2.dd))
+    }
+
+    /**
+      * Returns a representation of the formula `f1 == f2`.
+      */
+    override def mkXor(f1: BddFormula, f2: BddFormula): BddFormula = {
+      new BddFormula(creator.makeXor(f1.dd, f2.dd))
+    }
+
+    /**
+      * Returns the set of free variables in `f`.
+      */
+    override def freeVars(f: BddFormula): SortedSet[Int] = {
+      freeVarsAux(f.dd)
+    }
+
+    private def freeVarsAux(dd: DD): SortedSet[Int] = {
+      if (dd.isLeaf()) {
+        SortedSet.empty
+      } else {
+        SortedSet(dd.getVariable()) ++
+          freeVarsAux(dd.getLow) ++
+          freeVarsAux(dd.getHigh)
+      }
+    }
+
+    /**
+      * Applies the function `fn` to every variable in `f`.
+      */
+    override def map(f: BddFormula)(fn: Int => BddFormula): BddFormula = {
+      new BddFormula(mapAux(f.dd)(fn))
+    }
+
+    private def mapAux(dd: DD)(fn: Int => BddFormula): DD = {
+      if (dd.isLeaf()) {
+        dd
+      } else {
+        val currentVar = dd.getVariable()
+        val substDD = fn(currentVar).dd
+
+        val lowRes = mapAux(dd.getLow())(fn)
+        val highRes = mapAux(dd.getHigh())(fn)
+        creator.makeIte(substDD, highRes, lowRes)
+      }
+    }
+
+    /**
+      * Returns a representation equivalent to `f` (but potentially smaller).
+      */
+    override def minimize(f: BddFormula): BddFormula = f
+
+    /**
+      * Converts the given formula f into a type.
+      */
+    override def toType(f: BddFormula, env: Bimap[Symbol.KindedTypeVarSym, Int]): Type = {
+      createTypeFromBDDAux(f.dd, Type.True, env)
+    }
+
+    //TODO: Optimize
+    private def createTypeFromBDDAux(dd: DD, tpe: Type, env: Bimap[Symbol.KindedTypeVarSym, Int]): Type = {
+      if (dd.isLeaf()) {
+        return if (dd.isTrue()) tpe else Type.False
+      }
+
+      val currentVar = dd.getVariable()
+      val typeVar = env.getBackward(currentVar) match {
+        case Some(sym) => Type.Var(sym, SourceLocation.Unknown)
+        case None => throw InternalCompilerException(s"unexpected unknown ID: $currentVar")
+      }
+
+      val lowType = Type.mkApply(Type.And, List(tpe, Type.Apply(Type.Not, typeVar, SourceLocation.Unknown)), SourceLocation.Unknown)
```

Comment:
Here you use SourceLocation.Unknown, but this should be avoided at all costs (or at least when possible), because it means we dont give useful error messages. Can we use `tpe` or `env` to have better locations?